### PR TITLE
Fix incorrent field de-duplication

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

In a scenario like so:

    create table t1 (x int);
    create table t2 (x int);
    insert into t1 (x) values (1), (2), (3);
    insert into t2 (x) values (2), (3), (4);
    refresh table t1, t2;
    select * from (select * from t1 left join t2 on t1.x = t2.x) tjoin;

The query returned the value for `t1.x` twice because it got
de-duplicated due to the fields looking identical. Both had
`relation=tjoin` and `path=x`.

To address this, this changes the Field to contain the symbol the field
is pointing to. This way it includes a reference to the underlying
`t1.x` or `t2.x` `Reference` symbols and it correctly detects that the
fields on the upper level are different.

This should also make it easier to later on de-duplicate `fields()` and
`outputs()` on `AnalyzedRelation`

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/expression/symbol/Field.java [hunk 1]
```diff
@@ -60,15 +60,14 @@
 
     private final AnalyzedRelation relation;
     private final Path path;
-    private final DataType valueType;
+    private final Symbol pointer;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/Field.java]: Not sure what to call this - could also be `child`, `source`, `delegate` or just `symbol`.  Although `source` and `child` might be too overloaded as we also use it for the source relations of a relation.

I had also considered changing `Field` to `ScopedSymbol` to promote the fact that a `Field`s purpose is to put a symbol into a different scope. (Might follow up on this if others like this, but wouldn't do it as part of this PR as this is being backported)

File: sql/src/main/java/io/crate/analyze/QueriedSelectRelation.java [hunk 2]
```diff
@@ -143,11 +143,22 @@ public boolean hasAggregates() {
...
+     * but for this to work the new relation must have semantically equal outputs to the old relation.
+     */
     public AnalyzedRelation replaceSubRelation(AnalyzedRelation newSubRelation) {
         var mapFieldsToNewRelation = FieldReplacer.bind(
-            f -> f.relation().equals(subRelation)
-                ? newSubRelation.getField(f.path(), Operation.READ)
-                : f
+            f -> {
+                if (f.relation().equals(subRelation)) {
+                    int idx = subRelation.fields().indexOf(f);
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/QueriedSelectRelation.java]: shouldn't this use `newSubRelation`?

File: sql/src/test/java/io/crate/testing/SymbolMatchers.java [hunk 3]
```diff
@@ -135,17 +137,40 @@ protected Symbol featureValueOf(Symbol actual) {
...
+            @Override
+            protected T featureValueOf(Symbol actual) {
+                return getFeature.apply(actual);
+            }
+        };
+    }
+
     public static Matcher<Symbol> isReference(final String expectedName, @Nullable final DataType dataType) {
-        FeatureMatcher<Symbol, String> fm = new FeatureMatcher<Symbol, String>(equalTo(expectedName), "name", "name") {
+        FeatureMatcher<Symbol, String> fm = new FeatureMatcher<>(equalTo(expectedName), "name", "name") {
```
Comment (seut) [inline sql/src/test/java/io/crate/testing/SymbolMatchers.java]: could we use `withFeature()` here instead?

File: sql/src/main/java/io/crate/analyze/QueriedSelectRelation.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/QueriedSelectRelation.java]: It looks up the position of the old field in the old relation and returns the field at the same position in the new relation.

**Comment by mfussenegger** [comment]
ok to test

**Pull request merged**